pkg/fqdn/name_manager.go

```diff
@@ -274,6 +274,13 @@ func (n *NameManager) UpdateGenerateDNS(ctx context.Context, lookupTime time.Tim
 
 	g, ctx := errgroup.WithContext(ctx)
 	g.Go(func() error {
+		if ipcacheRevision == 0 {
+			log.WithFields(logrus.Fields{
+				"updatedDNSNames": updatedDNSNames,
+				"ipcacheRevision": ipcacheRevision,
+			}).Warn("FQDN: WaitForRevision ZERO")
```

Comment:
IIRC this is somewhat likely to happen when restating the agent on larger clusters.

The reason is that the ipcache does not start applying until all k8s resources are synchronized, which can take some time. However, we expect to be able to respond to DNS requests as soon as endpoints are restored, even if we can't allocate new IPs. We hope that, in this time period, we don't see any new IPs :-).

You can simulate this by adding a `time.Sleep()` in the codebase. I'm pretty sure this will happen, but there are a lot of moving parts and my mental model may be out-of-date.

Reply:
Are you saying that not waiting during restart is actually intentional? For the Envoy NetworkPolicy update we already skip pushing changes to Envoy during restart so that there is a chance for the internal resource cache to get filled up, hopefully to the same state that existed before restart. And since we are not actually sending anything to Envoy until the initial k8s resource sync is complete, the wait for the ACKs on those updates is also eliminated. This means that UseCurrent() does not put anything to the waitgroup during restart. That's the NetworkPolicy part, not sure about the ipcache itself. If we'd wait for a non-zero revision here during restart I guess DNS responses would stall during restart?

Reply:
As Casey said, IPCache can't inject new IPs (submitted via `UpsertLabels`) until K8s caches are synced. This is done so the IP identities (which need to collect labels from multiple K8s resources) don't flap during startup and that restored numeric identities are preserved. 

When I did the most recent FQDN refactor (introducing `fqdn` labels), I somehow didn't think about the case when there are new DNS responses are observed  before regeneration. So at least from my side, it was not intentional that we don't wait on IPCache. But even now that I'm aware of the issue, I'm not sure what the correct behavior would be. I see three options:

- Release the response without waiting on IPCache and hope IPCache is ready soon. This may lead to drops until K8s caches are synced (and seems to be the current behavior)
- The Wait for K8s resources to be synced and IPCache being ready. This may lead to DNS lookup timeouts
- Send back a DNS error response before IPCache is ready (e.g. SERVFAIL) and assume the client will retry once we're ready (which experience tells me won't happen and will lead to bug reports)

Reply:
I has a bit more of a think and realized that ipcache revision zero is fairly unlikely to happen except in the very early stages of startup. Because we very quickly start queuing changes, e.g. from reading Nodes.

What we do now is no different than if there was a policy stall: wait for 100ms then release the DNS response anyways. Then hope that TCP retries buy us enough time for the policy to be updated.

The easiest solution is probably for `GetPendingRevision()` to return `max(1, revision)`.

Reply:
Or could we skip the revision number `0` to start with?

Reply:
Wait, this is already the case:
https://github.com/cilium/cilium/blob/main/pkg/ipcache/metadata.go#L111